Added `register_multi_post_accumulate_grad_hook`

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #132610
* #129259
* __->__ #131949

This PR adds `torch.autograd.graph.register_multi_post_accumulate_grad_hook()` that registers a hook that runs after all passed-in leaf tensors' `AccumulateGrad` nodes. In other words, this is like an "all" hook over their `register_post_accumulate_grad_hook()`s.

```
def register_multi_post_accumulate_grad_hook(
    tensors: Sequence[torch.Tensor], fn: Callable[[Tuple[torch.Tensor, ...]], None]
) -> RemovableHandle:
```

The hook `fn` is expected to have the same signature as for `register_post_accumulate_grad_hook`:
```
def fn(tensor: torch.Tensor) -> None:
```

This function can be useful for distributed applications where we want to run some gradient reduction after a group of tensors' gradients have been computed (and accumulated).

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

Commit:
Added `register_multi_post_accumulate_grad_hook`

[ghstack-poisoned]

Commit:
Update on "Added `register_multi_post_accumulate_grad_hook`"




[ghstack-poisoned]

Commit:
Update on "Added `register_multi_post_accumulate_grad_hook`"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Commit:
Update on "Added `register_multi_post_accumulate_grad_hook`"


This PR adds `torch.autograd.graph.register_multi_post_accumulate_grad_hook()` that registers a hook that runs after all passed-in leaf tensors' `AccumulateGrad` nodes. In other words, this is like an "all" hook over their `register_post_accumulate_grad_hook()`s.

```
def register_multi_post_accumulate_grad_hook(
    tensors: Sequence[torch.Tensor], fn: Callable[[Tuple[torch.Tensor, ...]], None]
) -> RemovableHandle:
```

The hook `fn` is expected to have the same signature as for `register_post_accumulate_grad_hook`:
```
def fn(tensor: torch.Tensor) -> None:
```

This function can be useful for distributed applications where we want to run some gradient reduction after a group of tensors' gradients have been computed (and accumulated).

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

awgu commented on torch/autograd/graph.py:
```diff
@@ -547,6 +547,63 @@ def wrapped_fn(grad: torch.Tensor) -> None:
+    acc_grads = [_get_grad_fn_or_grad_acc(t) for t in tensors]
+    count: Dict[int, int] = dict()
+    nb_calls = None
+    lock = threading.Lock()
```
credit: https://github.com/pytorch/pytorch/pull/132055

soulitzer commented on torch/autograd/graph.py:
```diff
@@ -547,6 +547,63 @@ def wrapped_fn(grad: torch.Tensor) -> None:
+    The hook ``fn`` will get the ``tensors`` passed as :attr:`tensors`. The
+    hook is like an "all" version of :meth:`torch.Tensor.register_post_accumulate_grad_hook`,
+    where the hook only runs after each tensor's post-accumulate-grad hook
+    would have run.
```
```suggestion
    has run.
```

soulitzer commented on torch/autograd/graph.py:
```diff
@@ -547,6 +547,63 @@ def wrapped_fn(grad: torch.Tensor) -> None:
+        fn(tensors: Tuple[torch.Tensor, ...]) -> None
+
+    The hook ``fn`` will get the ``tensors`` passed as :attr:`tensors`. The
+    hook is like an "all" version of :meth:`torch.Tensor.register_post_accumulate_grad_hook`,
```
```suggestion
    hook is like an "all" multi-hook version of :meth:`torch.Tensor.register_post_accumulate_grad_hook`,
```

soulitzer reviewed (approved):
Thanks, looks great! Just some minor comments on the doc wording

Commit:
Update on "Added `register_multi_post_accumulate_grad_hook`"


This PR adds `torch.autograd.graph.register_multi_post_accumulate_grad_hook()` that registers a hook that runs after all passed-in leaf tensors' `AccumulateGrad` nodes. In other words, this is like an "all" hook over their `register_post_accumulate_grad_hook()`s.

```
def register_multi_post_accumulate_grad_hook(
    tensors: Sequence[torch.Tensor], fn: Callable[[Tuple[torch.Tensor, ...]], None]
) -> RemovableHandle:
```

The hook `fn` is expected to have the same signature as for `register_post_accumulate_grad_hook`:
```
def fn(tensor: torch.Tensor) -> None:
```

This function can be useful for distributed applications where we want to run some gradient reduction after a group of tensors' gradients have been computed (and accumulated).

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

sanketpurandare reviewed (approved):
Much awaited feature :)

awgu commented:
going to wait for the FSDP2 PR on top to fix the compile issue before landing this one

Commit:
Update on "Added `register_multi_post_accumulate_grad_hook`"


This PR adds `torch.autograd.graph.register_multi_post_accumulate_grad_hook()` that registers a hook that runs after all passed-in leaf tensors' `AccumulateGrad` nodes. In other words, this is like an "all" hook over their `register_post_accumulate_grad_hook()`s.

```
def register_multi_post_accumulate_grad_hook(
    tensors: Sequence[torch.Tensor], fn: Callable[[Tuple[torch.Tensor, ...]], None]
) -> RemovableHandle:
```

The hook `fn` is expected to have the same signature as for `register_post_accumulate_grad_hook`:
```
def fn(tensor: torch.Tensor) -> None:
```

This function can be useful for distributed applications where we want to run some gradient reduction after a group of tensors' gradients have been computed (and accumulated).

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

Commit:
Update on "Added `register_multi_post_accumulate_grad_hook`"


This PR adds `torch.autograd.graph.register_multi_post_accumulate_grad_hook()` that registers a hook that runs after all passed-in leaf tensors' `AccumulateGrad` nodes. In other words, this is like an "all" hook over their `register_post_accumulate_grad_hook()`s.

```
def register_multi_post_accumulate_grad_hook(
    tensors: Sequence[torch.Tensor], fn: Callable[[Tuple[torch.Tensor, ...]], None]
) -> RemovableHandle:
```

The hook `fn` is expected to have the same signature as for `register_post_accumulate_grad_hook`:
```
def fn(tensor: torch.Tensor) -> None:
```

This function can be useful for distributed applications where we want to run some gradient reduction after a group of tensors' gradients have been computed (and accumulated).

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

yf225 committed:
Update on "Added `register_multi_post_accumulate_grad_hook`"


This PR adds `torch.autograd.graph.register_multi_post_accumulate_grad_hook()` that registers a hook that runs after all passed-in leaf tensors' `AccumulateGrad` nodes. In other words, this is like an "all" hook over their `register_post_accumulate_grad_hook()`s.

```
def register_multi_post_accumulate_grad_hook(
    tensors: Sequence[torch.Tensor], fn: Callable[[Tuple[torch.Tensor, ...]], None]
) -> RemovableHandle:
```

The hook `fn` is expected to have the same signature as for `register_post_accumulate_grad_hook`:
```
def fn(tensor: torch.Tensor) -> None:
```

This function can be useful for distributed applications where we want to run some gradient reduction after a group of tensors' gradients have been computed (and accumulated).

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]
